KAFKA-17565: Move MetadataCache interface to metadata module

FrankYang0529 opened the pull request:
### Changes

* Move MetadataCache interface to metadata module and change Scala function to Java.
* Remove functions `getTopicPartitions`, `getAliveBrokers`, `topicNamesToIds`, `topicIdInfo`, and `getClusterMetadata` from MetadataCache interface, because these functions are only used in test code.

### Performance

* ReplicaFetcherThreadBenchmark
  ```
  ./jmh-benchmarks/jmh.sh -f 1 -i 2 -wi 2 org.apache.kafka.jmh.fetcher.ReplicaFetcherThreadBenchmark
  ```
  * trunk
  ```
  Benchmark                                  (partitionCount)  Mode  Cnt       Score   Error  Units
  ReplicaFetcherThreadBenchmark.testFetcher               100  avgt    2    4775.490          ns/op
  ReplicaFetcherThreadBenchmark.testFetcher               500  avgt    2   25730.790          ns/op
  ReplicaFetcherThreadBenchmark.testFetcher              1000  avgt    2   55334.206          ns/op
  ReplicaFetcherThreadBenchmark.testFetcher              5000  avgt    2  488427.547          ns/op
  ```
  * branch
  ```
  Benchmark                                  (partitionCount)  Mode  Cnt       Score   Error  Units
  ReplicaFetcherThreadBenchmark.testFetcher               100  avgt    2    4825.219          ns/op
  ReplicaFetcherThreadBenchmark.testFetcher               500  avgt    2   25985.662          ns/op
  ReplicaFetcherThreadBenchmark.testFetcher              1000  avgt    2   56056.005          ns/op
  ReplicaFetcherThreadBenchmark.testFetcher              5000  avgt    2  497138.573          ns/op
  ```

* KRaftMetadataRequestBenchmark
  ```
  ./jmh-benchmarks/jmh.sh -f 1 -i 2 -wi 2 org.apache.kafka.jmh.metadata.KRaftMetadataRequestBenchmark
  ```
  * trunk
  ```
  Benchmark                                                      (partitionCount)  (topicCount)  Mode  Cnt         Score   Error  Units
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                10           500  avgt    2    884933.558          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                10          1000  avgt    2   1910054.621          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                10          5000  avgt    2  21778869.337          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                20           500  avgt    2   1537550.670          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                20          1000  avgt    2   3168237.805          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                20          5000  avgt    2  29699652.466          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                50           500  avgt    2   3501483.852          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                50          1000  avgt    2   7405481.182          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                50          5000  avgt    2  55839670.124          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              10           500  avgt    2       333.667          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              10          1000  avgt    2       339.685          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              10          5000  avgt    2       334.293          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              20           500  avgt    2       329.899          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              20          1000  avgt    2       347.537          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              20          5000  avgt    2       332.781          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              50           500  avgt    2       327.085          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              50          1000  avgt    2       325.206          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              50          5000  avgt    2       316.758          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                10           500  avgt    2         7.569          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                10          1000  avgt    2         7.565          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                10          5000  avgt    2         7.574          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                20           500  avgt    2         7.568          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                20          1000  avgt    2         7.557          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                20          5000  avgt    2         7.585          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                50           500  avgt    2         7.560          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                50          1000  avgt    2         7.554          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                50          5000  avgt    2         7.574          ns/op
  ```
  * branch
  ```
  Benchmark                                                      (partitionCount)  (topicCount)  Mode  Cnt         Score   Error  Units
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                10           500  avgt    2    910337.770          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                10          1000  avgt    2   1902351.360          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                10          5000  avgt    2  22215893.338          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                20           500  avgt    2   1572683.875          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                20          1000  avgt    2   3188560.081          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                20          5000  avgt    2  29984751.632          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                50           500  avgt    2   3413567.549          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                50          1000  avgt    2   7303174.254          ns/op
  KRaftMetadataRequestBenchmark.testMetadataRequestForAllTopics                50          5000  avgt    2  54293721.640          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              10           500  avgt    2       318.335          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              10          1000  avgt    2       331.386          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              10          5000  avgt    2       332.944          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              20           500  avgt    2       340.322          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              20          1000  avgt    2       330.294          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              20          5000  avgt    2       342.154          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              50           500  avgt    2       341.053          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              50          1000  avgt    2       335.458          ns/op
  KRaftMetadataRequestBenchmark.testRequestToJson                              50          5000  avgt    2       322.050          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                10           500  avgt    2         7.538          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                10          1000  avgt    2         7.548          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                10          5000  avgt    2         7.545          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                20           500  avgt    2         7.597          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                20          1000  avgt    2         7.567          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                20          5000  avgt    2         7.558          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                50           500  avgt    2         7.559          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                50          1000  avgt    2         7.615          ns/op
  KRaftMetadataRequestBenchmark.testTopicIdInfo                                50          5000  avgt    2         7.562          ns/op
  ```

* PartitionMakeFollowerBenchmark
  ```
  ./jmh-benchmarks/jmh.sh -f 1 -i 2 -wi 2 org.apache.kafka.jmh.partition.PartitionMakeFollowerBenchmark
  ```
  * trunk
  ```
  Benchmark                                        Mode  Cnt    Score   Error  Units
  PartitionMakeFollowerBenchmark.testMakeFollower  avgt    2  158.816          ns/op
  ```
  * branch
  ```
  Benchmark                                        Mode  Cnt    Score   Error  Units
  PartitionMakeFollowerBenchmark.testMakeFollower  avgt    2  160.533          ns/op
  ```

* UpdateFollowerFetchStateBenchmark
  ```
  ./jmh-benchmarks/jmh.sh -f 1 -i 2 -wi 2 org.apache.kafka.jmh.partition.UpdateFollowerFetchStateBenchmark
  ```
  * trunk
  ```
  Benchmark                                                                Mode  Cnt     Score   Error  Units
  UpdateFollowerFetchStateBenchmark.updateFollowerFetchStateBench          avgt    2  4975.261          ns/op
  UpdateFollowerFetchStateBenchmark.updateFollowerFetchStateBenchNoChange  avgt    2  4880.880          ns/op
  ```
  * branch
  ```
  Benchmark                                                                Mode  Cnt     Score   Error  Units
  UpdateFollowerFetchStateBenchmark.updateFollowerFetchStateBench          avgt    2  5020.722          ns/op
  UpdateFollowerFetchStateBenchmark.updateFollowerFetchStateBenchNoChange  avgt    2  4878.855          ns/op
  ```


* CheckpointBench
  ```
  ./jmh-benchmarks/jmh.sh -f 1 -i 2 -wi 2 org.apache.kafka.jmh.server.CheckpointBench
  ```
  * trunk
  ```
  Benchmark                                         (numPartitions)  (numTopics)   Mode  Cnt  Score   Error   Units
  CheckpointBench.measureCheckpointHighWatermarks                 3          100  thrpt    2  0.997          ops/ms
  CheckpointBench.measureCheckpointHighWatermarks                 3         1000  thrpt    2  0.703          ops/ms
  CheckpointBench.measureCheckpointHighWatermarks                 3         2000  thrpt    2  0.486          ops/ms
  CheckpointBench.measureCheckpointLogStartOffsets                3          100  thrpt    2  1.038          ops/ms
  CheckpointBench.measureCheckpointLogStartOffsets                3         1000  thrpt    2  0.734          ops/ms
  CheckpointBench.measureCheckpointLogStartOffsets                3         2000  thrpt    2  0.637          ops/ms
  ```
  * branch
  ```
  Benchmark                                         (numPartitions)  (numTopics)   Mode  Cnt  Score   Error   Units
  CheckpointBench.measureCheckpointHighWatermarks                 3          100  thrpt    2  0.990          ops/ms
  CheckpointBench.measureCheckpointHighWatermarks                 3         1000  thrpt    2  0.659          ops/ms
  CheckpointBench.measureCheckpointHighWatermarks                 3         2000  thrpt    2  0.508          ops/ms
  CheckpointBench.measureCheckpointLogStartOffsets                3          100  thrpt    2  0.923          ops/ms
  CheckpointBench.measureCheckpointLogStartOffsets                3         1000  thrpt    2  0.736          ops/ms
  CheckpointBench.measureCheckpointLogStartOffsets                3         2000  thrpt    2  0.637          ops/ms
  ```

* PartitionCreationBench
  ```
  ./jmh-benchmarks/jmh.sh -f 1 -i 2 -wi 2 org.apache.kafka.jmh.server.PartitionCreationBench
  ```
  * trunk
  ```
  Benchmark                            (numPartitions)  (useTopicIds)  Mode  Cnt  Score   Error  Units
  PartitionCreationBench.makeFollower               20          false  avgt    2  5.997          ms/op
  PartitionCreationBench.makeFollower               20           true  avgt    2  6.961          ms/op
  ```
  * branch
  ```
  Benchmark                            (numPartitions)  (useTopicIds)  Mode  Cnt  Score   Error  Units
  PartitionCreationBench.makeFollower               20          false  avgt    2  6.212          ms/op
  PartitionCreationBench.makeFollower               20           true  avgt    2  7.005          ms/op
  ```

Reviewers: Ismael Juma <ismael@juma.me.uk>, David Arthur <mumrah@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

ijuma commented on core/src/main/scala/kafka/server/metadata/KRaftMetadataCache.scala:
```diff
@@ -345,68 +344,69 @@ class KRaftMetadataCache(
+  override def getAliveBrokerNodes(listenerName: ListenerName): util.List[Node] = {
     _currentImage.cluster().brokers().values().asScala.filterNot(_.fenced()).
-      flatMap(_.node(listenerName.value()).toScala).toSeq
+      flatMap(_.node(listenerName.value()).toScala).toList.asJava
```
Note that `toList` converts to an immutable linked list - you do _not_ want to do that if you are then calling `asJava`. `toBuffer` is probably a better option.

ijuma commented on core/src/main/scala/kafka/server/metadata/KRaftMetadataCache.scala:
```diff
@@ -345,68 +344,69 @@ class KRaftMetadataCache(
 
-  override def getAliveBrokerNodes(listenerName: ListenerName): Seq[Node] = {
+  override def getAliveBrokerNodes(listenerName: ListenerName): util.List[Node] = {
     _currentImage.cluster().brokers().values().asScala.filterNot(_.fenced()).
```
We should remove `asScala` from here and use `stream` instead.

FrankYang0529 replied:
Thanks for the suggestion. I will update it tomorrow.

ijuma commented:
I realize your PR is still wip, so feel free to ignore the two comments if you were still working through that code. :)

chia7712 commented:
@FrankYang0529 please fix the conflicts, thanks

FrankYang0529 committed:
KAFKA-17565: Move MetadataCache interface to metadata module

Signed-off-by: PoAn Yang <payang@apache.org>

mumrah commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -821,17 +821,19 @@ object TestUtils extends Logging {
+          broker.metadataCache.numPartitions(topic).isEmpty
         } else {
-          broker.metadataCache.numPartitions(topic) == Some(expectedNumPartitions)
+          broker.metadataCache.numPartitions(topic).isPresent && broker.metadataCache.numPartitions(topic).get == expectedNumPartitions
```
I think this can be simplified to use `filter` or maybe `orElse(null) = `

mumrah commented on core/src/main/scala/kafka/server/metadata/KRaftMetadataCache.scala:
```diff
@@ -345,68 +350,75 @@ class KRaftMetadataCache(
-      flatMap(_.node(listenerName.value()).toScala)
+  override def getAliveBrokerNode(brokerId: Int, listenerName: ListenerName): util.Optional[Node] = {
+    util.Optional.ofNullable(_currentImage.cluster.broker(brokerId))
+      .filter(broker => !broker.fenced)
```
With Java 11 we can now use `Predicate.not` for negating filters with lambda. This can be

```
.filter(Predicate.not(_.fenced))
```

mumrah commented on core/src/main/scala/kafka/server/metadata/KRaftMetadataCache.scala:
```diff
@@ -320,18 +320,23 @@ class KRaftMetadataCache(
 
-  override def getTopicId(topicName: String): Uuid = _currentImage.topics().topicsByName().asScala.get(topicName).map(_.id()).getOrElse(Uuid.ZERO_UUID)
+  override def getTopicId(topicName: String): Uuid = util.Optional.ofNullable(_currentImage.topics.topicsByName.get(topicName))
+    .map(topic => topic.id)
```
We can use the method reference style here: `.map(_.id)`

mumrah commented on core/src/main/scala/kafka/server/metadata/KRaftMetadataCache.scala:
```diff
@@ -239,12 +239,12 @@ class KRaftMetadataCache(
+                                errorUnavailableListeners: Boolean = false): util.List[MetadataResponseTopic] = {
     val image = _currentImage
-    topics.toSeq.flatMap { topic =>
+    topics.asScala.toSeq.flatMap { topic =>
```
Can we avoid this conversion to Scala and back to Java collection? Since `topics` is now a Java set, it seems like we can use a Java stream here.

mumrah commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2383,7 +2385,7 @@ class KafkaApis(val requestChannel: RequestChannel,
         val brokers = new DescribeClusterResponseData.DescribeClusterBrokerCollection()
         val describeClusterRequest = request.body[DescribeClusterRequest]
-        metadataCache.getBrokerNodes(request.context.listenerName).foreach { node =>
+        metadataCache.getBrokerNodes(request.context.listenerName).asScala.foreach { node =>
```
Do we need to convert to Scala for the `foreach` here? Could we use the Java stream and its `forEach` instead?

mumrah commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1193,8 +1195,8 @@ class KafkaApis(val requestChannel: RequestChannel,
             .filter(_.leaderId != MetadataResponse.NO_LEADER_ID)
-            .flatMap(metadata => metadataCache.
-                getAliveBrokerNode(metadata.leaderId, request.context.listenerName))
+            .flatMap(metadata => OptionConverters.toScala(metadataCache.
```
Can we let `topicMetadata` stay as as a Java collection and convert this whole stream to a Java stream?

mumrah commented on core/src/main/scala/kafka/server/DelayedElectLeader.scala:
```diff
@@ -74,7 +75,7 @@ class DelayedElectLeader(
     val metadataCache = replicaManager.metadataCache
     val completedPartitions = waitingPartitions.collect {
-      case (tp, leader) if metadataCache.getLeaderAndIsr(tp.topic, tp.partition).exists(_.leader == leader) => tp
+      case (tp, leader) if OptionConverters.toScala(metadataCache.getLeaderAndIsr(tp.topic, tp.partition)).exists(_.leader == leader) => tp
```
The `exists` can be done in Java with `filter` + `isPresent`

mumrah commented on core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala:
```diff
@@ -186,9 +188,9 @@ class AddPartitionsToTxnManager(
   private def getTransactionCoordinator(partition: Int): Option[Node] = {
-   metadataCache.getLeaderAndIsr(Topic.TRANSACTION_STATE_TOPIC_NAME, partition)
+   OptionConverters.toScala(metadataCache.getLeaderAndIsr(Topic.TRANSACTION_STATE_TOPIC_NAME, partition))
       .filter(_.leader != MetadataResponse.NO_LEADER_ID)
-      .flatMap(metadata => metadataCache.getAliveBrokerNode(metadata.leader, interBrokerListenerName))
+      .flatMap(metadata => OptionConverters.toScala(metadataCache.getAliveBrokerNode(metadata.leader, interBrokerListenerName)))
```
I think this would be better if we keep the Java Optional and convert it at the end.

mumrah commented on core/src/main/scala/kafka/cluster/Replica.scala:
```diff
@@ -113,7 +113,7 @@ class Replica(val brokerId: Int, val topicPartition: TopicPartition, val metadat
       val cachedBrokerEpoch = metadataCache.getAliveBrokerEpoch(brokerId)
       // Fence the update if it provides a stale broker epoch.
-      if (brokerEpoch != -1 && cachedBrokerEpoch.exists(_ > brokerEpoch)) {
+      if (brokerEpoch != -1 && cachedBrokerEpoch.isPresent && cachedBrokerEpoch.get > brokerEpoch) {
```
Simplify to `filter` + `isPresent`

mumrah reviewed (commented):
@FrankYang0529, thanks for this patch! I'm glad to see MetadataCache moving to Java :) 

I left several minor comments inline. One overall comment is that I think it would be best to include the parens for no-arg function calls in the Scala code (only for Scala code already included/modified in this PR)

Since MetadataCache is in the hot path for many requests, it would make sense to run our benchmarks before/after this patch to make sure we don't introduce a performance regression.

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 committed:
address comment

Signed-off-by: PoAn Yang <payang@apache.org>

FrankYang0529 commented:
Hi @mumrah, thanks for the review. I addressed all comments. For benchmark, I think we can do it in https://issues.apache.org/jira/browse/KAFKA-17573 when we migrate `KRaftMetadataCache`. The `MetadataCache` in this PR is more about interface, not implementation.

mumrah commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1192,17 +1192,18 @@ class KafkaApis(val requestChannel: RequestChannel,
             .filter(_.leaderId != MetadataResponse.NO_LEADER_ID)
-            .flatMap(metadata => OptionConverters.toScala(metadataCache.
-                getAliveBrokerNode(metadata.leaderId, request.context.listenerName)))
+            .map(metadata => metadataCache.getAliveBrokerNode(metadata.leaderId, request.context.listenerName))
+            .filter(_.isPresent)
+            .map(_.get)
```
I think we could reduce this to a flatMap with Optional#stream

mumrah reviewed (commented):
@FrankYang0529 the changes since my last review look good. One minor thing inline.

Can you fix the conflicts?

As for the benchmark, I'm still interested in seeing it for this PR since we are touching so many critical areas of the broker. Specifically, I want to make sure we're not inadvertently copying collections or creating excessive objects. A benchmark should reveal such issues.

mumrah commented:
I ran one of the benchmarks and there actually seems to be a slight improvement! cc @ijuma 

benchmark:
```
./jmh.sh -f 2 -i 2 -wi 2 org.apache.kafka.jmh.server.CheckpointBench.measureCheckpointHighWatermarks
```

trunk:
```
Benchmark                                        (numPartitions)  (numTopics)   Mode  Cnt  Score   Error   Units
CheckpointBench.measureCheckpointHighWatermarks                3          100  thrpt    4  0.154 ± 0.010  ops/ms
CheckpointBench.measureCheckpointHighWatermarks                3         1000  thrpt    4  0.101 ± 0.032  ops/ms
CheckpointBench.measureCheckpointHighWatermarks                3         2000  thrpt    4  0.091 ± 0.019  ops/ms
JMH benchmarks done
```

this PR:
```
Benchmark                                        (numPartitions)  (numTopics)   Mode  Cnt  Score   Error   Units
CheckpointBench.measureCheckpointHighWatermarks                3          100  thrpt    4  0.153 ± 0.109  ops/ms
CheckpointBench.measureCheckpointHighWatermarks                3         1000  thrpt    4  0.141 ± 0.031  ops/ms
CheckpointBench.measureCheckpointHighWatermarks                3         2000  thrpt    4  0.129 ± 0.033  ops/ms
```

@FrankYang0529 now I definitely think you should run the benchmarks for this PR so you can take credit for the performance improvement 😁

FrankYang0529 commented:
@mumrah, thanks for the suggestion and review. I will do benchmark and fix the conflict tomorrow. Appreciate it!

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 committed:
address comment

Signed-off-by: PoAn Yang <payang@apache.org>

mumrah reviewed (commented):
@FrankYang0529 Thanks for the updates. This looks good to me!

I would still like to see to the relevant benchmark results with this PR so we can include them in the commit message.

FrankYang0529 commented:
@mumrah I update benchmark to PR description. The result is a little different from yours. The difference between trunk and branch results on my laptop is close, so I think the result is acceptable.

chia7712 commented on core/src/main/scala/kafka/server/metadata/KRaftMetadataCache.scala:
```diff
@@ -345,68 +352,75 @@ class KRaftMetadataCache(
-    image.cluster().brokers().values().asScala.filterNot(_.fenced()).
-      map(b => new BrokerMetadata(b.id, b.rack))
+  private def getAliveBrokers(image: MetadataImage): util.List[BrokerMetadata] = {
+    _currentImage.cluster().brokers().values().stream()
```
`_currentImage` -> `image`

chia7712 commented on metadata/src/main/java/org/apache/kafka/metadata/MetadataCache.java:
```diff
@@ -0,0 +1,243 @@
+     */
+    Optional<Integer> numPartitions(String topic);
+
+    Map<String, Uuid> topicNamesToIds();
```
it is useless - maybe we can remove it to simplify the interface.

chia7712 commented on metadata/src/main/java/org/apache/kafka/metadata/MetadataCache.java:
```diff
@@ -0,0 +1,243 @@
+
+    Map<Uuid, String> topicIdsToNames();
+
+    Map.Entry<Map<String, Uuid>, Map<Uuid, String>> topicIdInfo();
```
ditto

chia7712 commented on metadata/src/main/java/org/apache/kafka/metadata/MetadataCache.java:
```diff
@@ -0,0 +1,243 @@
+
+    Map<Integer, Node> getPartitionReplicaEndpoints(TopicPartition tp, ListenerName listenerName);
+
+    Cluster getClusterMetadata(String clusterId, ListenerName listenerName);
```
ditto

chia7712 commented on core/src/main/scala/kafka/server/metadata/KRaftMetadataCache.scala:
```diff
@@ -345,68 +352,75 @@ class KRaftMetadataCache(
+  override def getPartitionReplicaEndpoints(tp: TopicPartition, listenerName: ListenerName): util.Map[Integer, Node] = {
     val image = _currentImage
-    val result = new mutable.HashMap[Int, Node]()
+    val result = new mutable.HashMap[Integer, Node]()
```
could you please use java Map instead?

chia7712 commented on core/src/main/scala/kafka/server/metadata/KRaftMetadataCache.scala:
```diff
@@ -239,35 +239,37 @@ class KRaftMetadataCache(
+            .setTopicId(Option(image.topics().getTopic(topic).id()).getOrElse(Uuid.ZERO_UUID))
+            .setIsInternal(Topic.isInternal(topic))
+            .setPartitions(partitionMetadata.toBuffer.asJava))
+        case None =>  util.stream.Stream.empty()
```
please remove the redundant space

chia7712 commented on core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala:
```diff
@@ -186,9 +188,9 @@ class AddPartitionsToTxnManager(
   }
 
   private def getTransactionCoordinator(partition: Int): Option[Node] = {
```
we can use Java optional as it has only one usage

chia7712 commented on test-common/test-common-runtime/src/main/java/org/apache/kafka/common/test/ClusterInstance.java:
```diff
@@ -290,7 +290,7 @@ default void waitForTopic(String topic, int partitions) throws InterruptedExcept
             () -> brokers.stream().allMatch(broker -> partitions == 0 ?
                 broker.metadataCache().numPartitions(topic).isEmpty() :
-                broker.metadataCache().numPartitions(topic).contains(partitions)
+                broker.metadataCache().numPartitions(topic).get() == partitions
```
`broker.metadataCache().numPartitions(topic).filter(p -> p == partitions).isPresent()`

chia7712 commented on core/src/main/scala/kafka/server/metadata/KRaftMetadataCache.scala:
```diff
@@ -239,35 +239,37 @@ class KRaftMetadataCache(
     val result = new DescribeTopicPartitionsResponseData()
     breakable {
-      topics.foreach { topicName =>
+      topics.asScala.foreach { topicName =>
```
`forEachRemaining`

chia7712 reviewed (commented):
@FrankYang0529 thanks for this patch. a couple of comments are left. PTAL

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 committed:
address comment

Signed-off-by: PoAn Yang <payang@apache.org>

FrankYang0529 committed:
remove useless functions

Signed-off-by: PoAn Yang <payang@apache.org>

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 committed:
ci(AutoTopicCreationManagerTest): fix error

Signed-off-by: PoAn Yang <payang@apache.org>

chia7712 commented on metadata/src/main/java/org/apache/kafka/metadata/MetadataCache.java:
```diff
@@ -0,0 +1,237 @@
+
+    boolean hasAliveBroker(int brokerId);
+
+    List<BrokerMetadata> getAliveBrokers();
```
this is used by test infra only, and we can rewrite the check by `hasAliveBroker`. 

*before*
```
        TestUtils.waitForCondition(() ->
                brokers().values().stream().allMatch(brokerServer -> brokerServer.metadataCache().getAliveBrokers().size() == brokers.size()),
            "Failed to wait for publisher to publish the metadata update to each broker.");
```

*after*
```
        TestUtils.waitForCondition(() ->
                brokers.values().stream().map(BrokerServer::metadataCache)
                    .allMatch(cache -> brokers.values().stream().map(b -> b.config().brokerId()).allMatch(cache::hasAliveBroker)),
            "Failed to wait for publisher to publish the metadata update to each broker.");
```

chia7712 commented on metadata/src/main/java/org/apache/kafka/metadata/MetadataCache.java:
```diff
@@ -0,0 +1,237 @@
+
+    Set<String> getAllTopics();
+
+    Set<TopicPartition> getTopicPartitions(String topicName);
```
it seems the usage of this method is to get the number of partitions. could you please rewrite the tests to use `numPartitions` instead?

chia7712 reviewed (commented):
@FrankYang0529 thanks for your patch. I leave some comments to cleanup the interface

FrankYang0529 committed:
address comment

Signed-off-by: PoAn Yang <payang@apache.org>

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

mumrah committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 committed:
fix ci error

Signed-off-by: PoAn Yang <payang@apache.org>

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 commented:
Hi @chia7712, I addressed all comments. Could you review this when you have time? Thanks.

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

chia7712 commented:
@FrankYang0529 Could you please briefly summarize the changes of this PR? Also, it would be better to list the deleted (unused) APIs

FrankYang0529 commented:
@chia7712 I update PR description. Thanks.

chia7712 commented on metadata/src/main/java/org/apache/kafka/metadata/MetadataCache.java:
```diff
@@ -0,0 +1,233 @@
+        }
+    }
+
+    private static List<BrokerMetadata> getAliveBrokers(MetadataImage image) {
```
Do we really need to use `BrokerMetadata`? It seems we can remove `getAliveBrokers` and then simplify `getRandomAliveBroker`.
```java
    private static Optional<Integer> getRandomAliveBroker(MetadataImage image) {
        var aliveBrokers = image.cluster().brokers().values().stream()
                .filter(broker -> !broker.fenced())
                .map(BrokerRegistration::id)
                .toList();
        if (aliveBrokers.isEmpty()) return Optional.empty();
        return Optional.of(aliveBrokers.get(ThreadLocalRandom.current().nextInt(aliveBrokers.size())));
    }
```

FrankYang0529 replied:
Update it. Thanks.

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-17565

FrankYang0529 committed:
address comment

Signed-off-by: PoAn Yang <payang@apache.org>

chia7712 reviewed (approved):
LGTM
